File: src/main/java/dev/ipparser/gui/LogConfig.java
package dev.ipparser.gui;

import dev.ipparser.storage.AppDb;

/**
 * All "log settings" state that used to live as a pile of {@code volatile}
 * fields in the window class. Read by worker threads during a scan; written on
 * the EDT from the Log Settings dialog. Persisted to SQLite via {@link AppDb}.
 *
 * Changes are applied by mutating this object on the EDT; because each field is
 * a plain volatile reference read consistently enough for single-field checks,
 * the scanner threads observe whole-before-during-start values. Short-lived
 * races during a live scan are acceptable and were also present before.
 */
public final class LogConfig {

    // ----- what gets logged -----
    public volatile boolean logOpen = true;
    public volatile boolean logClosed;
    public volatile boolean logTimeout;
    public volatile boolean logError;
    public volatile boolean logActions = true;
    public volatile boolean external; // status 2 toggle

    // ----- MC-probe filters -----
    public final McFilters.Settings mc = new McFilters.Settings();

    private LogConfig() {
    }

    public static LogConfig load() {
        LogConfig c = new LogConfig();
        AppDb db = AppDb.get();
        c.logOpen = db.getBool("log.open", c.logOpen);
        c.logClosed = db.getBool("log.closed", c.logClosed);
        c.logTimeout = db.getBool("log.timeout", c.logTimeout);
        c.logError = db.getBool("log.error", c.logError);
        c.logActions = db.getBool("log.actions", c.logActions);
        c.external = db.getBool("syntax.external", c.external);

        c.mc.online = db.getBool("filter.online", c.mc.online);
        c.mc.onlineOp = db.get("filter.onlineOp", c.mc.onlineOp);
        c.mc.onlineVal = db.get("filter.onlineVal", c.mc.onlineVal);

        c.mc.version = db.getBool("filter.version", c.mc.version);
        c.mc.versionOp = db.get("filter.versionOp", c.mc.versionOp);
        c.mc.versionVal = db.get("filter.versionVal", c.mc.versionVal);

        c.mc.brand = db.getBool("filter.brand", c.mc.brand);
        c.mc.brandVal = db.get("filter.brandVal", c.mc.brandVal);

        c.mc.motd = db.getBool("filter.motd", c.mc.motd);
        c.mc.motdVal = db.get("filter.motdVal", c.mc.motdVal);

        c.mc.players = db.getBool("filter.players", c.mc.players);
        c.mc.playersVal = db.get("filter.playersVal", c.mc.playersVal);
        return c;
    }

    public void save() {
        AppDb.Adapter a = AppDb.get().adapter();
        a.setBool("log.open", logOpen);
        a.setBool("log.closed", logClosed);
        a.setBool("log.timeout", logTimeout);
        a.setBool("log.error", logError);
        a.setBool("log.actions", logActions);
        a.setBool("syntax.external", external);

        a.setBool("filter.online", mc.online);
        a.set("filter.onlineOp", mc.onlineOp);
        a.set("filter.onlineVal", mc.onlineVal);

        a.setBool("filter.version", mc.version);
        a.set("filter.versionOp", mc.versionOp);
        a.set("filter.versionVal", mc.versionVal);

        a.setBool("filter.brand", mc.brand);
        a.set("filter.brandVal", mc.brandVal);

        a.setBool("filter.motd", mc.motd);
        a.set("filter.motdVal", mc.motdVal);

        a.setBool("filter.players", mc.players);
        a.set("filter.playersVal", mc.playersVal);
    }
}

File: src/log_config.rs
//! Port of the Java `LogConfig`: all "log settings" state that used to live in
//! the window class. Read by worker threads during a scan; written from the
//! Log Settings dialog. Persisted to SQLite via [`crate::storage`].

use crate::filters::FilterSettings;
use crate::storage as db;

/// Log/filter settings group.
#[derive(Debug, Clone)]
pub struct LogConfig {
    // ----- what gets logged -----
    pub log_open: bool,
    pub log_closed: bool,
    pub log_timeout: bool,
    pub log_error: bool,
    pub log_actions: bool,
    pub external: bool, // status 2 toggle

    // ----- MC-probe filters -----
    pub mc: FilterSettings,
}

impl Default for LogConfig {
    fn default() -> Self {
        Self {
            log_open: true,
            log_closed: false,
            log_timeout: false,
            log_error: false,
            log_actions: true,
            external: false,
            mc: FilterSettings::with_defaults(),
        }
    }
}

impl LogConfig {
    /// Loads the config from the SQLite settings store.
    pub fn load() -> Self {
        let d = Self::default();
        Self {
            log_open: db::get_bool("log.open", d.log_open),
            log_closed: db::get_bool("log.closed", d.log_closed),
            log_timeout: db::get_bool("log.timeout", d.log_timeout),
            log_error: db::get_bool("log.error", d.log_error),
            log_actions: db::get_bool("log.actions", d.log_actions),
            external: db::get_bool("syntax.external", d.external),
            mc: FilterSettings {
                online: db::get_bool("filter.online", false),
                online_op: db::get("filter.onlineOp", "is"),
                online_val: db::get("filter.onlineVal", ""),

                version: db::get_bool("filter.version", false),
                version_op: db::get("filter.versionOp", "is"),
                version_val: db::get("filter.versionVal", ""),

                brand: db::get_bool("filter.brand", false),
                brand_val: db::get("filter.brandVal", ""),

                motd: db::get_bool("filter.motd", false),
                motd_val: db::get("filter.motdVal", ""),

                players: db::get_bool("filter.players", false),
                players_val: db::get("filter.playersVal", ""),
            },
        }
    }

    /// Persists the config to the SQLite settings store.
    pub fn save(&self) {
        db::set_bool("log.open", self.log_open);
        db::set_bool("log.closed", self.log_closed);
        db::set_bool("log.timeout", self.log_timeout);
        db::set_bool("log.error", self.log_error);
        db::set_bool("log.actions", self.log_actions);
        db::set_bool("syntax.external", self.external);

        db::set_bool("filter.online", self.mc.online);
        db::set("filter.onlineOp", &self.mc.online_op);
        db::set("filter.onlineVal", &self.mc.online_val);

        db::set_bool("filter.version", self.mc.version);
        db::set("filter.versionOp", &self.mc.version_op);
        db::set("filter.versionVal", &self.mc.version_val);

        db::set_bool("filter.brand", self.mc.brand);
        db::set("filter.brandVal", &self.mc.brand_val);

        db::set_bool("filter.motd", self.mc.motd);
        db::set("filter.motdVal", &self.mc.motd_val);

        db::set_bool("filter.players", self.mc.players);
        db::set("filter.playersVal", &self.mc.players_val);
    }
}
